blas64/blas64.go

```diff
@@ -0,0 +1,331 @@
+)
+
+var blas64 blas.Float64 = goblas.Blas{}
+
+// Use sets the BLAS float64 implementation to be used by subsequent BLAS calls.
+// The default implementation is goblas.Blas.
+func Use(b blas.Float64) {
+	blas64 = b
+}
+
+// Implementation returns the current BLAS float64 implementation.
+//
+// Implementation allows direct calls to the current the BLAS float64 implementation
+// giving finer control of parameters.
+func Implementation() blas.Float64 {
+	return blas64
+}
+
+// Vector represents a vector with an associated element increment.
+type Vector struct {
+	Inc  int
+	Data []float64
+}
+
+// General represents a matrix using the conventional storage scheme.
+type General struct {
+	Rows, Cols int
+	Stride     int
+	Data       []float64
+}
+
+// Band represents a band matrix using the band storage scheme.
+type Band struct {
+	Rows, Cols int
+	KL, KU     int
+	Stride     int
+	Data       []float64
+}
+
+// Triangular represents a triangular matrix using the conventional storage scheme.
+type Triangular struct {
+	N      int
+	Stride int
+	Data   []float64
+	Uplo   blas.Uplo
+	Diag   blas.Diag
+}
+
+// TriangularBand represents a triangular matrix using the band storage scheme.
+type TriangularBand struct {
+	N, K   int
+	Stride int
+	Data   []float64
+	Uplo   blas.Uplo
+	Diag   blas.Diag
+}
+
+// TriangularPacked represents a triangular matrix using the packed storage scheme.
+type TriangularPacked struct {
+	N    int
+	Data []float64
+	Uplo blas.Uplo
+	Diag blas.Diag
+}
+
+// Symmetric represents a symmetric matrix using the conventional storage scheme.
+type Symmetric struct {
+	N      int
+	Stride int
+	Data   []float64
+	Uplo   blas.Uplo
+}
+
+// SymmetricBand represents a symmetric matrix using the band storage scheme.
+type SymmetricBand struct {
+	N, K   int
+	Stride int
+	Data   []float64
+	Uplo   blas.Uplo
+}
+
+// SymmetricPacked represents a symmetric matrix using the packed storage scheme.
+type SymmetricPacked struct {
+	N    int
+	Data []float64
+	Uplo blas.Uplo
+}
+
+// Level 1
+
+const (
+	negIncX = "blas64: negative x increment"
+	negIncY = "blas64: negative y increment"
+)
+
+// Dot will panic if either vector increment is negative.
+func Dot(n int, x, y Vector) float64 {
+	if x.Inc < 0 {
+		panic(negIncX)
+	}
+	if y.Inc < 0 {
+		panic(negIncY)
+	}
+	return blas64.Ddot(n, x.Data, x.Inc, y.Data, y.Inc)
+}
+
+// Nrm2 will panic if the vector increment is negative.
+func Nrm2(n int, x Vector) float64 {
+	if x.Inc < 0 {
+		panic(negIncX)
+	}
+	return blas64.Dnrm2(n, x.Data, x.Inc)
+}
+
+// Asum will panic if the vector increment is negative.
+func Asum(n int, x Vector) float64 {
+	if x.Inc < 0 {
+		panic(negIncX)
+	}
+	return blas64.Dasum(n, x.Data, x.Inc)
+}
+
+// Iamax will panic if the vector increment is negative.
+func Iamax(n int, x Vector) int {
+	if x.Inc < 0 {
+		panic(negIncX)
+	}
+	return blas64.Idamax(n, x.Data, x.Inc)
+}
+
+// Swap will panic if either vector increment is negative.
```

Comment:
Swap, copy, axpy, etc panics are incorrect. With negative vectors they do what you'd expect,  and I assume they are useful given they are in blas. It's the single vector cases that have surprising behavior with a negative increment, and this should panic.